### DISCUSSION TITLE
Electrow HMI integration

### DISCUSSION THREAD

**Discussion by JCDS71** [emsesp/EMS-ESP32#1676] [General]
Hello everybody.
I would like to talk about a project to use an HMI screen to control the EMS-ESP bus with a single ESP32 touch display (Electrow: https://www.elecrow.com/esp-terminal-with-esp32-3-5 -inch-parallel-480x320-tft-capacitive-touch-display-rgb-by-chip-ili9488.html).
I managed to use the REST API from my display, via wifi,  by querying an ESP32 connected to the bus. This way I can control the boiler / thermostat through this touch screen and display the data in real time.
From now on, I would like to merge the 2 codes into a single ESP32 application which should, in theory, have sufficient resources to operate.
I'am looking for an entry point in the code that will allow me to read and write data from the EMS bus. @Proddy suggested creating a Stub Class (I don't know exactly what that consists of but I'll do some research...) and I have to say that it's very complicated to understand the architecture of the code because I'm almost a beginner in C++, but I'm hanging in there :).
I welcome any help and indications that could help me continue to develop this project.
THANK YOU

**Comment by proddy** [comment]
Share the code on GitHub and I'll take a look on what the best approach is to integrate this into the main EMS-ESP core.

**Comment by JCDS71** [comment]
Hello Proddy, I've just uploaded the mixed project (HMI + EMS-ESP) : https://github.com/JCDS71/EMS-ESP32-3.6.4_HMI
There is just the display initialization that is implemented.

"src/display.cpp" ("common_Display.h") is used to initialize the TFT display and touch screen.
Most of the code that manages the graphic's interface (based on LVGL 8.4.0) is in the "src/UI" directory.

Let me know if you need more informations. This code is based on examples coming from ELECROW and were based on Arduino Framework, but not platformio, so I had to "convert" it manually. Later I will try to optimize everything because I'm aware that there is a lot of files that don't need to be compiled :)

Regards

**Reply by proddy** [reply]
Hi @JCDS71 - I had a quick look at the code. I noticed you're calling the HTTP RestFul API service in EMS-ESP using a new HTTP client. A better way could be is call it direct from the loop, similar to how we use it in `emsesp::publish_all_loop()`, with a timer and send directly to `lv_obj_add_state()` so removing the extra HTTP client and also avoid blocking code in the thread that could crash the ESP32.

**Reply by JCDS71** [reply]
Hi @proddy,
Thank you for the informations. I'm looking for how to retrieve the values, but analyzing the code seems very complicated to me, especially since there is almost no documentation.
I understand that I have to place the "Display.loop()" loop in "emsesp::publish_all_loop()" but I don't understand how to retrieve the data, for example from the thermostat.
I see the function "EMSESP::get_device_value_info()" but how do I call it from my "Display.cpp" file, what parameters should I provide to it?
Also, I would like to know how to use a debug console to display information like Serial.print() would?
If you could provide me with some piece of code, I would be grateful.
Regards

**Reply by proddy** [reply]
It'll would be too complicated to explain how the code works if you have only a fundamental concept of C++ and object-orientated programming. Have a look a the class EMSdevice which is inherited by every EMS device (like a thermostat). You can add a function similar to generate_values() which exposes all the data you need into a JSON document, which can be called from the main class. Then I would traverse this json object array and send the commands to the HMI interface as needed.
